### ISSUE TITLE
constants as part of a linear expression?

### ISSUE THREAD

**Issue by Cellophil** [PyPSA/linopy#96]
More of a question than a feature request, possibly it's complicated:

Linear expressions are usually a sum over variables with prefactors plus a constant. That constant needs to go on the RHS in the usual constraint formulation, that's probably the reason why the linearExpr object does not support constants. However, if I wanted make use of the nice algebra of the linearExpr object for, say, "1-x", I have to separate off the constant early. In power systems I might want to express something like "20% of headroom" = 0.2 (p_nom-p).

My hacky solution would be to create a neutral variable=1 early on, and work with that, that should be fine. Could that be an option overall, say referencing a variable index -42 or something, and when creating a constraint having a logic sort it over to the RHS?

**Comment by FabianHofmann** [comment]
Hey @Cellophil I love the creativity but it seems a bit hacky. I would rather add a data field to the LinearExpression data called `constant` here https://github.com/PyPSA/linopy/blob/360536e1f5d23761274d8bc6440ffea2d9746853/linopy/expressions.py#L160 

And whenever adding numeric values to an expression, it would add to the `constant` data field. The question is a bit how to store the data in model? Or even more concrete: At which point do we separate move the numeric values to the rhs and the variables to the lhs?

**Comment by Cellophil** [comment]
That definitely sounds ... sound, if that's a straight-forward option I would love it. I guess the magic function where <=, ==, >= is evaluated and the constraint object is created would be a good spot to separate?
Subtracting and adding numbers would then probably require radd and rsub.
I don't know the ins and outs of the code yet though. Will need some reading :-)

**Comment by FabianHofmann** [comment]
Yes, you are probably right. So I'd suggest to create a function to tidy up the sides, like:

```python 
def sanitize_constraint_sides(lhs, rhs):
    """
    Adjust the two sides of a constraint.   

    This function is used to ensure that numerical values are always on the
    right hand side of a constraint. This is needed when two expressions with 
    constants are used to define a constraint.
    """
    def separate_constant(side):
        if isinstance(side, expressions.LinearExpression):
            return side - side.constant,  side.constant
        elif isinstance(side, variables.Variable):
            return side.to_linexpr(), 0
        else:
            return 0, DataArray(side)

    lhs, lhs_constant = separate_constant(lhs)
    rhs, rhs_constant = separate_constant(rhs)
    
    # we might want to align the two sides here as well. 
    return lhs - rhs, rhs_constant - lhs_constant
```

**Comment by FabianHofmann** [comment]
this is totally non-tested, but might give you a start.

**Comment by Cellophil** [comment]
Hi,

I checked it out a bit, couple of design questions...
- should it be possible to set up a linear expression *only* with a constant - I would almost say yes, "1" is a valid linear expression in the math sense, but it would require changes to the constructor
- should "const" be a required and always present field in a linear expression - or optional

I wonder whether treating a constant as a variable with a special index is so hacky after all, it would pull through the current logic quite nicely possibly? In a way "const" introduces a special field that needs branching and additional logic everywhere, while a special index needs special treatment exactly in le, ge, eq. Introducing const as an extra field is possibly not such a small addition?

**Comment by FabianHofmann** [comment]
Hey @Cellophil, good questions. 
> should it be possible to set up a linear expression only with a constant - I would almost say yes, "1" is a valid linear expression in the math sense, but it would require changes to the constructor

I would say yes, but not expect that to be the ordinary use case. Constants would be swallowed by linear expressions once they are added/substracted 

> should "const" be a required and always present field in a linear expression - or optional
Always present but can well be empty.

> I wonder whether treating a constant as a variable with a special index is so hacky after all, it would pull through the current logic quite nicely possibly? In a way "const" introduces a special field that needs branching and additional logic everywhere, while a special index needs special treatment exactly in le, ge, eq. Introducing const as an extra field is possibly not such a small addition?

This could well be. I tend to like the idea, after thinking about it. The only thing that felt distorting is to have a placeholder in the variable data fields for terms which do not have a variable. But we also keep masked entries there. 
And with this approach, things like 

1 + x - 4y 

Would actually lead to 3 (sub-)terms, which make sense. And things like 

5 - y + 17 + x

would lead to four terms, which one would see in the representation of the linear expression just as is, meaning no automated summation of the constants. That would happen in the sanitization. 

The only thing is that LinearExpression are always attached to a model instance. So when creating a LinearExpression from a constant one needs to make sure that the model is passed. Something like 

```python
    @classmethod
    def from_constant(arr, model):
        """
        Create a linear expression from an array of constant value.

        The resulting expression is a linear expression which has coefficients of the
        passed constant and variables of -42 as a placeholder. 

        Parameters
        ----------
        arr : array_like
        model : linopy.Model
        """
        arr = as_dataarray(arr)
        variable_placeholder = xr.full_like(arr, -42)
        ds = Dataset({"coeffs": arr, "vars": variable_placeholder})
        return LinearExpression(ds, model)
```

and then a function that ensure the type conversion in `__add__` / `__sub__` and possibly others, like

```python 

def maybe_to_expression(ds, model):
    "Returns a LinearExpression if argument is not a LinearExpression already."
    if isinstance(ds, LinearExpression):
        return ds
    elif isinstance(ds, variables.Variable):
        return LinearExpression.from_tuples((1, ds))
    else:
        # assume ds is a array like object that represents a constant
        return LinearExpression.from_constant(ds, model)
```

But this is just guessing. Might be more tricky than that.

Would you want to give it a try?

**Comment by lumbric** [comment]
I think this is a duplicate of #60. Since this ticket contains already quite some discussion and helpful hints, you might want to close #60 as duplicate.

**Comment by Cellophil** [comment]
Sorry I'm running a delay... My attempt: #101
